sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/Distinct.java

```diff
@@ -78,26 +83,41 @@
    */
   public static <T, IdT> WithRepresentativeValues<T, IdT> withRepresentativeValueFn(
       SerializableFunction<T, IdT> fn) {
-    return new WithRepresentativeValues<T, IdT>(fn, null);
+    return new WithRepresentativeValues<>(fn, null);
   }
 
   @Override
   public PCollection<T> expand(PCollection<T> in) {
-    return in
-        .apply("CreateIndex", MapElements.via(new SimpleFunction<T, KV<T, Void>>() {
-          @Override
-          public KV<T, Void> apply(T element) {
-            return KV.of(element, (Void) null);
-          }
-        }))
-        .apply(Combine.<T, Void>perKey(
-            new SerializableFunction<Iterable<Void>, Void>() {
+    if (!(in.getWindowingStrategy().getWindowFn() instanceof NonMergingWindowFn)) {
+      LOG.warn("Distinct is not currently guaranteed to remove all duplicates when using merging"
```

Comment:
If it is using the default trigger and zero allowed lateness, then it should work, right? Because in that case, a window is always GC'd after its one and only firing, so all merges happen before the first pane index. That is probably a common case. I think we should let that through silently and actually fully reject (aka `throw UnsupportedOperationException`) cases where there might be dupes.

At a minimum, I would have a builder method that says `.withPossibleDuplicatesInMergingWindows()` to explicitly acknowledge the bad behavior and toggle it on.